### DISCUSSION TITLE
Should Scan API be redesigned?

### DISCUSSION THREAD

**Discussion by axa88** [dotnet-bluetooth-le/dotnet-bluetooth-le#856] [Q&A]
Why does `StartScanningForDevicesAsync` not return until the scan has completed?
In that case shouldn't the name be `ScanForDevicesAsync`?

In the current implementation it is easy to fall out of sync with the `IsScanning` property as a call to `StopScanningForDevicesAsync` calls for the cancellation of `StartScanningForDevicesAsync` but returns as complete before that cancel is even processed.
Therefore to avoid that issue the work around would be to create 2 methods one to Start using the `StartScanningForDevicesAsync` method. And the other to Stop the scan, but not by calling `StopScanningForDevicesAsync` but to cancel the `CancelationToken` used to start the scan.

Also passing `StartScanningForDevicesAsync` a `CancellationToken` to disregard it in favor for the `Adapter.ScanTimeout` parameter, is just wanky and confusing.  On the one hand you are forced to use the `CancellationToken` to properly Stop the scan properly, and on the other hand your forced to use the separate timeout parameter to set the scan duration.  Its just a confusing design.
Then ultimately you need to use the `CancellationToken` to manually stopping the scan as indicated above, and use a second `CancleationToken` to detect if the scan itself failed.

**Comment by odussault05** [comment]
> Why does StartScanningForDevicesAsync not return until the scan has completed?
> In that case shouldn't the name be ScanForDevicesAsync?
When is it complete? When it founds the first device, what if I have more that one Bluetooth device nearby?

For the "janky" behaviour of the CancellationToken, this is how it is implemented in C#. It is not synchronous, the method needs to check if the token is cancelled to throw an exception.

**Reply by axa88** [reply]
> For the "janky" behaviour of the CancellationToken, this is how it is implemented in C#. It is not synchronous, the method needs to check if the token is cancelled to throw an exception.

what?
